# 字符串

python转移字符, 例如\t是制表符, 下面是一些常用的转义字符

\\    反斜杠

\'    单引号

\"    双引号

\a    响铃

\b    退格, (也就是删去最后一个字符)

\n    换行

\v    纵向制表符

\t    横向制表符

\r    回车

\f    换页

In [1]:
print('Hello world\'')

Hello world'


In [2]:
print('Hello world\a')

Hello world


In [3]:
print('Hellow world\b')

Hellow world


In [4]:
print('Hello world\nabc')

Hello world
abc


In [5]:
print('Hello world\tabc')

Hello world	abc


In [6]:
print('Hello world\rabc')

abc


In [7]:
print('Hello world\f abc')

Hello world abc


有时候我们并不希望输出转义的字符串, 那么就需要使用r来原始化字符串

In [8]:
print(r'\n\t\r\b\a')

\n\t\r\b\a


python 字符串格式化

In [9]:
print('My name is %s and weight is %d kg'%('zara', 56))

My name is zara and weight is 56 kg


%c	 格式化字符及其ASCII码

%s	 格式化字符串

%d	 格式化整数

%u	 格式化无符号整型

%o	 格式化无符号八进制数

%x	 格式化无符号十六进制数

%X	 格式化无符号十六进制数（大写）

%f	 格式化浮点数字，可指定小数点后的精度

%e	 用科学计数法格式化浮点数

%E	 作用同%e，用科学计数法格式化浮点数

%g	 %f和%e的简写

%G	 %f 和 %E 的简写

%p	 用十六进制数格式化变量的地址

# Python format 格式化函数

In [10]:
print('{} {}'.format('Hellow', 'world'))

Hellow world


In [11]:
print('{0} {1}'.format('Hello', 'world'))

Hello world


In [12]:
print('{0} {1} {0} {2}'.format('Hello', 'world', 'ocean'))

Hello world Hello ocean


format函数也可以设置参数

In [13]:
print('web site :{name}, address {url}'.format(name='bird', url='www.runoob.com'))

web site :bird, address www.runoob.com


通过字典设置参数

In [14]:
site = {'name':'bird', 'url':'www.runoob.com'}
print('web site:{name}, address:{url}'.format(**site))

web site:bird, address:www.runoob.com


# 数字格式化

In [16]:
print('{:.2f}'.format(3.141592653))

3.14


数字        格式       输出            描述

3.1415926  {:.2f}    3.14        保留小数点后两位

3.1415926  {:+.2f}  +3.14        带符号保留小数点后两位

-1         {:+.2f}  -1.00        带符号保留小数点后两位

2.71828    {:.0f}    3           不带小数

5          {:0>2d}   05          数字补零 (填充左边, 宽度为2)

5          {:x<4d}   5xxx        数字补x (填充右边, 宽度为4)

10         {:x<4d}   10xx        数字补x (填充右边, 宽度为4)

1000000    {:,}      1,000,000   以逗号分隔的数字格式

0.25       {:.2%}    25.00%      百分比格式

1000000000 {:.2e}    1.00e+09    指数记法

13         {:10d}         13     右对齐 (默认, 宽度为10)

13         {:<10d}    13         左对齐 (宽度为10)

13         {:^10d}      13       中间对齐 (宽度为10)

In [17]:
print('{:3.2f}'.format(1003.141592653))

1003.14


In [19]:
print('{:2.4f}'.format(3.14159265354))

3.1416


In [20]:
print('{:+2.3f}'.format(3.1415926))

+3.142


In [21]:
print('{:.0f}'.format(3.1415926))

3


In [24]:
print('{:0>2d}'.format(5)) # 少于两个宽度在左边补齐， 否则就原版输出

05


In [25]:
print('{:0<4d}'.format(32))# 少于两个宽度在右边补齐， 否则就原版输出

3200


In [26]:
print('{:,}'.format(100000000))

100,000,000


In [27]:
print('{:.2%}'.format(1.2456)) # 保留小数点后两位的百分制记法

124.56%


In [28]:
print('{:.2e}'.format(245243934693406)) # 保留小数点后两位的科学计数法记法

2.45e+14


In [29]:
print('{:10d}'.format(13)) # 右对齐， 默认宽度为10

        13


In [30]:
print('{:<10d}'.format(13)) # 左对齐, 默认10个宽度

13        


In [32]:
print('{:^10d}'.format(13)) # 中间对齐, 默认10个宽度

    13    


In [33]:
print("{} 对应的位置是 {{0}}".format("runoob")) # 我我们可以使用大括号来作为转义字符

runoob 对应的位置是 {0}


Python的字符串的内建函数

In [34]:
a = 'helloworld'

In [35]:
a1 = a.capitalize()# 将第一个字符大写
a2 = a.center(20,'*')

In [36]:
print('a1:', a1, 'a2:', a2)

a1: Helloworld a2: *****helloworld*****


In [38]:
a3 = a.count('o',0,len(a)) # 统计字符串中的字母出现的次数
print(a3)

2


In [39]:
print(a)
b = a.encode('ascii') # 用ascii编码
print(b)
c = b.decode('ascii')# 解码
print(c)

helloworld
b'helloworld'
helloworld


In [40]:
a.endswith('d',0,len(a)) # 检测是否以某个字符结尾

True

In [41]:
#把字符串 string 中的 tab 符号转为空格，tab 符号默认的空格数是 8
'afdsa\tasfdvdfb'.expandtabs()

'afdsa   asfdvdfb'

In [42]:
#检测 str 是否包含在 string 中，如果 beg 和 end 指定范围，则检查是否包含在指定范围内，
#如果是返回开始的索引值，否则返回-1
a.find('o',0,len(a)) # 只返回第一个的索引

4

In [43]:
a.index('h') # 如果不存在会报错

0

In [44]:
#如果 string 至少有一个字符并且所有字符都是字母或数字则返
#回 True,否则返回 False
a.isalpha()

True

In [45]:
a.isdigit()#只包含数字

False

In [46]:
a.isdecimal() # 只包含十进制数

False

In [47]:
a.islower() # 存在区分大小写的字符，而且所有的这些都是小写才返回true
#类似的有isupper, istitle

True

In [48]:
a.isnumeric()# 只包含数字字符

False

在此特别说明isdigit, isnumeric, isdecimal三种方法的差别

In [50]:
s1 = '１1'# 这里两个1分别是Unicode编码和全角的, 都可以判断为true
print(s1.isdigit())
print(s1.isnumeric())
print(s1.isdecimal())

True
True
True


In [51]:
s2 = b'1'
print(s2.isdigit()) 
# true
print(s2.isnumeric())
#AttributeError: 'bytes' object has no attribute 'isnumeric'
print(s2.isdecimal())
#AttributeError: 'bytes' object has no attribute 'isnumeric'

True


AttributeError: 'bytes' object has no attribute 'isnumeric'

In [53]:
num = "IV" # 罗马数字
print(num.isdigit())   # True
print(num.isdecimal()) # False
print(num.isnumeric()) # True
# 本机没有罗马数字输入，很尴尬

False
False
False


In [54]:
num = "四" # 汉字
print(num.isdigit())   # False
print(num.isdecimal()) # False
print(num.isnumeric()) # True

False
False
True


isdigit()

True: Unicode数字，byte数字（单字节），全角数字（双字节），罗马数字

False: 汉字数字

Error: 无


isdecimal()

True: Unicode数字，，全角数字（双字节）

False: 罗马数字，汉字数字

Error: byte数字（单字节）


isnumeric()

True: Unicode数字，全角数字（双字节），罗马数字，汉字数字

False: 无

Error: byte数字（单字节）

In [55]:
a = ' asfsdag '
a.lstrip()

'asfsdag '

In [56]:
a.rstrip()

' asfsdag'

In [57]:
a = 'asfd-sfad-asdf'
a.split('-')

['asfd', 'sfad', 'asdf']

In [58]:
s="hello world!<[www.google.com]>byebye"

In [59]:
s1 = s.split('[')[1]
s2 = s1.split(']')[0]
print(s2)

www.google.com


In [60]:
s2.split('.')

['www', 'google', 'com']

In [61]:
s2.split('.',1)

['www', 'google.com']